File: src/main/java/com/blogly/blogly/domain/user/Password.java
package com.blogly.blogly.domain.user;

import com.blogly.blogly.domain.exception.InvalidPasswordException;

import java.util.Objects;

public record Password(String hashedValue) {

    public static Password create(String plainText, PasswordHasher hasher) {
        Objects.requireNonNull(hasher, "Hasher cannot be null");

        if (plainText == null || plainText.length() < 8) {
            throw new InvalidPasswordException("Password must be at least 8 characters");
        }

        return new Password(hasher.hash(plainText));
    }

    public boolean matches(String plainText, PasswordHasher hasher) {
        return hasher.matches(plainText, this.hashedValue);
    }
}

File: src/main/java/com/blogly/blogly/domain/user/Password.kt
package com.blogly.blogly.domain.user

@JvmInline
value class Password(val hashedValue: String) {

    fun matches(plainText: String, hasher: PasswordHasher) = hasher.matches(plainText, hashedValue)

    companion object {
        private const val MIN_LENGTH = 8

        fun create(plainText: String, hasher: PasswordHasher): Password {
            require(plainText.length >= MIN_LENGTH) { "Password must be at least $MIN_LENGTH characters" }

            return Password(hasher.hash(plainText))
        }
    }
}
